Add Additional Instance Names to Info Pop-Over

colinmford opened the pull request:
Hey @LettError and @typemytype,

This is how it's looking so far... I added the style map stuff in there first — if you want me to remove the PostScript name from the main table and add it to this dialog I can certainly try to do that next... but because that was a change in behavior and it'll be a little more complicated than the style map names, I thought I would run this part by you guys first. 

Personally I might prefer to leave PostScript Names where it is in the main table, but if you really want it out of there I can try to make it work.

<img width="886" alt="image" src="https://github.com/LettError/designSpaceRoboFontExtension/assets/671409/01bfd74f-4c49-42d8-a781-3186401eb8ab">

I wanted it to match the RF Info panel as close as possible.

One potential issue with the radio buttons is you can't unset them once you select them, meaning you can't change your mind and leave that field blank if you wanted to. Potential solutions include adding the "use default value" checkboxes RF has, adding a "None" option to the radio button, or changing the input to a combo box or something. I welcome your input!

@typemytype also mentions in #91 the potential for auto-generating names, which is a behavior for PostScript names currently but I worry about getting too clever when it comes to the Style Map names. What do you think?

typemytype commented on DesignspaceEditor2.roboFontExt/lib/designspaceEditor/ui.py:
```diff
@@ -558,15 +558,38 @@ class InstancesAttributesPopover(BaseAttributePopover):
+        self.instanceStyleMapNames.container.textStyleMapStyleName = vanilla.TextBox((col1, y, col2-padding, 22), "Style Map Style Name:", alignment="right")
+        self.instanceStyleMapNames.container.editorStyleMapStyleName = vanilla.RadioGroup((col2, y, -10, 88), ["Regular", "Italic", "Bold", "Bold Italic"], callback=self.controlEditCallback)
+        
+        self.instanceStyleMapNames.container.editorStyleMapFamilyName.set(getattr(self.instancesDescriptor, "styleMapFamilyName", ""))
```
I guess an instanceDescriptor always has an attribute `styleMapFamilyName` and `styleMapStyleName`

it will be `None` by default. 

use `self.instanceDescriptor. styleMapFamilyName or ""`

colinmford replied:
I see, got it

colinmford commented:
Ok, I think I'm set with the style map part of it. @LettError @typemytype do you want me to go further and add the postscript name field to that dialog, removing it from the main table or should we call it done here?

LettError commented:
There's a contextual menu item for assembling the postscriptfontname from the available family and stylenames. Would it be useful to extend that into "do all names" or something like that?

colinmford commented:
@LettError @typemytype 

Check out this latest commit:

<img width="1171" alt="image" src="https://github.com/LettError/designSpaceRoboFontExtension/assets/671409/dda2c57c-1638-481e-b39f-08fbbb4b1333">

<img width="1074" alt="image" src="https://github.com/LettError/designSpaceRoboFontExtension/assets/671409/9a3b7b6f-533e-47ca-8ad9-5cab726146a5">


- Moved PostScript to the info pop over, removed it from the main table. Removed the "auto Postscript name" item from the context menu.
- Added two columns to 1) indicate that there are additional names and 2) indicate there are localized names, respectively (the source tab had the second one, it was missing from the instance tab so I thought I would add it)
- Added a field for the identifying "name" attribute... not typically necessary but for a sense of completeness, now all the attributes of the `<instance>` tag are covered in one way or another
- Added "magic" buttons to those fields for generating "best guess" names. It uses transformer functions added to `tools.py`
  - The postScript auto function more closely follows the spec in terms of not allowing certain glyphs
  - the style map function has a little bit of heuristics that I can probably guess will never cover 100% of cases, but hopefully does well enough for most cases
  - The identifying name function could probably do more to generate a "unique" name if there are more than one entry of the same name but that would require looking at all the instances in the document and I was hoping not to go _that far_... it just smushes family name and style name together
- I added a "help" button to that additional names tab in the popover, which links to the DS5 spec with more info about those names, but for some reason it doesn't work 100% of the time and sometimes it crashes RF if you click it too many times... and then I realized the other help button in the toolbar ALSO does not work really well, maybe @typemytype has some ideas...

typemytype commented on DesignspaceEditor2.roboFontExt/lib/designspaceEditor/ui.py:
```diff
@@ -558,16 +558,61 @@ class InstancesAttributesPopover(BaseAttributePopover):
+        # Create the text box and edit box for identifier Name
+        self.instanceStyleMapNames.container.identifierNameLabel = vanilla.TextBox((col1, y, col2-padding, 22), "Identifying Name:", alignment="right")
+        self.instanceStyleMapNames.container.identifierNameTextBox = vanilla.EditText((col2, y, -40, 22), "", callback=self.controlEditCallback)
+        self.instanceStyleMapNames.container.identifierNameAutoBtn = vanilla.SquareButton((-32, y, -10, 22), "􀜍", callback=self.autoIdentifyingNameCallback)
```
that unicode value will only work on systems where SF Symbols is installed and preferable this has to be an image from an SF symbol, see https://github.com/LettError/designSpaceRoboFontExtension/blob/e3c94c796cac46b1152c57bcf1f3b967ec0b0c52/DesignspaceEditor2.roboFontExt/lib/designspaceEditor/ui.py#L778 and add the SF Symbol as an asset in the [resource](https://github.com/LettError/designSpaceRoboFontExtension/blob/e3c94c796cac46b1152c57bcf1f3b967ec0b0c52/DesignspaceEditor2.roboFontExt/resources/toolbar_30_30_smallcircle.filled.circle.fill.pdf) folder

colinmford replied:
Done.

LettError commented:
Screenshot looks great! I don’t have time until Friday to see it run. But I’m supporting the work!Doc’s wlll need an update too. ErikOn 5 Jun 2024, at 22:29, Colin M. Ford ***@***.***> wrote:﻿
@colinmford commented on this pull request.



In DesignspaceEditor2.roboFontExt/lib/designspaceEditor/ui.py:
>  
+        # Create the container for the Style Map tab
+        self.instanceStyleMapNames.container = vanilla.Box((10, 10, -10, -10))
+        self.styleMapStyleOptions = ["regular", "italic", "bold", "bold italic"]
+
+        col1 = 10
+        col2 = 180
+        padding = 24
+        y = 10
+
+        # Create the text box and edit box for identifier Name
+        self.instanceStyleMapNames.container.identifierNameLabel = vanilla.TextBox((col1, y, col2-padding, 22), "Identifying Name:", alignment="right")
+        self.instanceStyleMapNames.container.identifierNameTextBox = vanilla.EditText((col2, y, -40, 22), "", callback=self.controlEditCallback)
+        self.instanceStyleMapNames.container.identifierNameAutoBtn = vanilla.SquareButton((-32, y, -10, 22), "􀜍", callback=self.autoIdentifyingNameCallback)

Done.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

colinmford commented:
@LettError I can update it... is it all written straight into the HTML in `DesignspaceEditor2.roboFontExt/html/index.html` or is that generated somehow?

colinmford commented:
While we're in here, I have a sort of partially related question — should the "update UFO filename" function also use the postscript name transformer function I wrote? 

At the moment if you have a font with the name "Polymath Text" "Bold Italic", it'll create a name like `instances/Polymath Text-Bold Italic.ufo`, while the PostScript name it'll come up with is `PolymathText-BoldItalic`... it might be a cleaner look if they both follow the same method.

ryanbugden commented:
Would it be possible to choose an auto-formatter in prefs?
- Postscript Style
- Spaces
- Underscores

colinmford commented:
@LettError @typemytype 

OK, I updated the doc page in `DesignspaceEditor2.roboFontExt/html/index.html` and changedthe "update UFO filename" function also use the postscript name transformer function I wrote.

LMK what you think!

LettError commented:
Github is confusing, I can't find the page with the changes to the docs? Or also how to get a working extension with the PR code.

colinmford commented:
> Github is confusing, I can't find the page with the changes to the docs? Or also how to get a working extension with the PR code.

In a pinch you could just [get the zip from here my repo here.](https://github.com/colinmford/designSpaceRoboFontExtension/archive/refs/heads/feature-styleMapInstances.zip)

colinmford commented on DesignspaceEditor2.roboFontExt/lib/designspaceEditor/tools.py:
```diff
@@ -196,3 +197,51 @@ def getObjectValue_forString_errorDescription_(self, value, string, error):
+
+        return filtered_name
+    
+    # TODO: spec says max length is 63, but should we check this
```
@typemytype yeah but just doing it crudely could result in something like `AVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryLongNameHere-` which since it ends with a hyphen and doesn't include a style name might cause some application somewhere to complain... though it is unlikely...

I made the function a little more smart, it'll take from the family name up to a point then start taking from the style name until it reaches the limit... I guess it doesn't cover every situation but 63 characters is actually pretty long!

typemytype replied:
super smart!! thanks

LettError commented:
Installed it. Looking at it. Sparkly icon is a bit dim in a dark RF.
<img width="160" alt="image" src="https://github.com/LettError/designSpaceRoboFontExtension/assets/1243961/5f70eb51-fde0-45b9-991a-18d3d5659951">

LettError commented:
Seems to work as advertised.

colinmford commented:
> Installed it. Looking at it. Sparkly icon is a bit dim in a dark RF. <img alt="image" width="160" src="https://private-user-images.githubusercontent.com/1243961/339974539-5f70eb51-fde0-45b9-991a-18d3d5659951.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg2MzgxNjEsIm5iZiI6MTcxODYzNzg2MSwicGF0aCI6Ii8xMjQzOTYxLzMzOTk3NDUzOS01ZjcwZWI1MS1mZGUwLTQ1YjktOTkxYS0xOGQzZDU2NTk5NTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTdUMTUyNDIxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTA1NDk0OGJiNzQ0MjgxYmY2OTAxY2Y2NWUzYjk2MTIwNWNlM2RhY2U0ZTY0OWJiM2FmYThiMTUyYjM3NzIzYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.S7iZuNBlmR1bBM12_0fb4mOXS_XpiLfTKhFK6A6xQOE">

@typemytype, is there a way to provide darkmode versions of the icons, or at least to get the darkmode state of RF?

LettError commented:
Colin can you elaborate on this name. Is it new? What is its purpose?
<img width="878" alt="image" src="https://github.com/LettError/designSpaceRoboFontExtension/assets/1243961/6fe68792-104f-4023-a86c-d5a71e64a1d4">

typemytype commented:
use `symbolImage("wand.and.stars", "primary")` to use one of the dynamic OS colors

colinmford commented:
@typemytype Thanks! Working now

<img width="290" alt="image" src="https://github.com/LettError/designSpaceRoboFontExtension/assets/671409/f833d3c3-3a2b-4ed3-8439-3fbb466e45bf">

typemytype commented:
is the  `instance.name` necessary for the DSE? where is this used? I guess this should look more like an identifier then a readable string

colinmford commented:
I've opened up a question in FontTools https://github.com/fonttools/fonttools/issues/3572 about how unique `instance.name` really needs to be... I looked around in a few libraries and it doesn't really seem to be used.

But to answer the question of why I included it — with its inclusion we would now support all the attributes of the `<instance>` tag. We can certainly remove it if you do not want it there, I just added it for completeness.

colinmford commented:
@LettError @typemytype How about this — I removed the "identifying name" field from the pop over, and added a line to automatically name instances like is done for Sources (i.e. "instance.##") if the `name` attribute is missing. 

<img width="907" alt="image" src="https://github.com/LettError/designSpaceRoboFontExtension/assets/671409/9cff879b-87c2-4123-a62e-e8fc7ba35054">

https://github.com/colinmford/designSpaceRoboFontExtension/blob/069de2016b363494cbc50017258666487a2e5cbd/DesignspaceEditor2.roboFontExt/lib/designspaceEditor/ui.py#L148-L149

LettError commented:
I'm in the middle of TM exams and a bit pressed for time. I understand you also have your schedules. My current thinking: both the sourceDescriptor.name and the instanceDescriptor.name are for tools that need to identify these records. Whether or not this is still a common practice I don't know. The sourceDescriptor.name is probably used (what was the related issue?). I doubt it will be necessary or possible to change the DS5 spec. 
I'd argue to just keep the names out of the editor. It's not been necessary to edit these values by hand.

colinmford commented:
@LettError Sure, no problem we can leave this until after exams

typemytype commented:
DSE doesnt use instance.name anywhere (and as you found out no other tool needs this either) so I would leave this like it is. If tools needs this they can always fill in the name attribute.
 
FYI: DSE has to review source names which are important in DSE to identify sources ([see](https://github.com/LettError/designSpaceRoboFontExtension/issues/93))

colinmford commented:
@typemytype yeah I agree 100%. 

To me, this last commit is good enough to fulfill the purposes of this PR. I won't bug you guys any more with additions, but if there's any further changes you guys want made after you have space and time to review, I will certainly do it. Otherwise, when you're ready, I would support merging this so we can call it done.

typemytype commented:
Look good to me!

typemytype commented:
thanks Colin, super cool addition to DSE!

colinmford commented:
Thanks @typemytype and @LettError !

ryanbugden commented:
This is great, thank you @colinmford and team! 

Is there a select-all-instances-and-wave-magic-wand option?

typemytype commented:
no, the popup info is only for the selected instance.
I guess those magic-wand-options could be menu options too applied to all selected instances

ryanbugden commented:
I think that would be super useful, like the Update UFO Filename and Update PostScript Font Name were!
